### ISSUE TITLE
l2 regularization for bias too.. is it necessary?

### ISSUE THREAD

**Issue by nbansal90** [ikhlestov/vision_networks#21]
Thank you for your effort, it is really helping me in my project. Illarion, I had a small doubt in the implementation part.

While you have applied l2_regularization, you have applied to all the parameters ie weights and bias. Is that advisory? given that l2 is mainly/mostly applied only to weights.

l2_loss = tf.add_n( [tf.nn.l2_loss(var) for var in tf.trainable_variables()])

Shouldn't we just add the tf.nn.l2_loss for weights only ?

**Comment by ikhlestov** [comment]
Hi!
Mainly, as I know, "applying weight decay to the bias units usually makes only a small difference to the final network". So I didn't drop the just for simplicity. You may try to train network without bias term regularization, and reopen the issue with the results.

**Comment by nbansal90** [comment]
Thanks Illarion for your prompt reply!
Its not an issue as such, but I was trying to implement it in a different way and getting "large difference" in terms of "regularization loss". What I did was to print the regularization loss achieved after each epoch of training, using the way you have done. and the result which i got was:
** ------------------------------ Train epoch: 17 ------------------------------

Training...
mean cross_entropy: 0.351586, mean l2_loss: 4122.128742, mean accuracy: 0.879021
Validation...
mean cross_entropy: 0.629531, mean accuracy: 0.798377
Time per epoch: 0:01:58.125252, Est. complete in: 9:17:09
**

I made a small change to the code, where I passed the kernel regularizer as a parameter to output = tf.layers.conv2d(_input, out_features, kernel_size, strides=(1,1), padding='SAME',
                                kernel_initializer=self.orthogonal_initializer(),kernel_regularizer=self.l2_reg())
and then collecting loss as:
        reg_ws = tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSSES)
        reg_loss = tf.reduce_sum(reg_ws)
        self.reg_loss = reg_loss

gave l2_loss as:
** ------------------------------ Train epoch: 17 ------------------------------

Training...
mean cross_entropy: 0.351586, mean l2_loss: 0.128742, mean accuracy: 0.879021
Validation...
mean cross_entropy: 0.629531, mean accuracy: 0.798377
Time per epoch: 0:01:58.125252, Est. complete in: 9:17:09
**

which is a significant change. I am not able to understand the reason for this. What would you suggest/comment on this?

Thank you.
